**Experiment No. 04**
---
**Aim:** Installing jupyter Notebook and unserstanding Sklearn library.

---

**Objectives:**
* To install jupyter notebook.
* To explore datasets in the sklearn library.
* To understand and implement various Machine Learning algorithms using Sk learn library.

---

**Datasets:** Breast Cancer Dataset of sklearn library.

---

**Theory:**

Scikit-learn is a free software machine learning library for the python programming language.It features various classification, regression and clustering algorithms including support vector machines, random forests, gradient bboosting, k-means and DBSCAN and iss designed to interoperate with Python numerical and scientific libraries using NumPy and SciPy.

Scikit-learn is largely written in Python, and uses numpy extensively for high-performance linear algebra and array operations. Furthermore, some core algorithms are written in Cython to improve performance. Support vector machines are implemented by a Cython wrapper around LIBSVM; logistic regression and linear support vector machines by a similar wrapper around LIBLINEAR. In such cases, extending these methods with Python may not be possible.

Scikit-learn integrates well with many other Python libraries, such as matplotlib and plotly for plotting, numpy for array vectorization, pandas dataframes, scipy, and many more.

---

In this experiment, dataset classifies the cancer tumor into malignant or a benign tumor depending ont the various input parameters.
This experiment uses Support Vector Classifier to classify the data into any of the predefined classes.

![alt text](https://drive.google.com/file/d/17N10Dr9GLffWZCGADhxC8X9gYoR-s5jN/view?usp=sharing)

![alt text](https://https://drive.google.com/drive/u/1/folders/1VTt2eI3xiLcvfTTHrd2KOZbQlmZU2Rkq)

![alt text](https://https://drive.google.com/drive/u/1/folders/1VTt2eI3xiLcvfTTHrd2KOZbQlmZU2Rkq)


Link To the installation process:https://drive.google.com/open?id=1zupFpRfeqWqULjqMENg5NctyWO5gotWn


The following cell imports all the libraaries needed to implement the experiment

In [0]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split

In [0]:
cancer  = datasets.load_breast_cancer()
df = pd.DataFrame(data=cancer.data ,columns = cancer.feature_names )
print(df.head(2))
print(df.info())


   mean radius  mean texture  ...  worst symmetry  worst fractal dimension
0        17.99         10.38  ...          0.4601                  0.11890
1        20.57         17.77  ...          0.2750                  0.08902

[2 rows x 30 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
mean radius                569 non-null float64
mean texture               569 non-null float64
mean perimeter             569 non-null float64
mean area                  569 non-null float64
mean smoothness            569 non-null float64
mean compactness           569 non-null float64
mean concavity             569 non-null float64
mean concave points        569 non-null float64
mean symmetry              569 non-null float64
mean fractal dimension     569 non-null float64
radius error               569 non-null float64
texture error              569 non-null float64
perimeter error            569 non-null float64
area error             

The above code loads the breast cancer dataset from the sklearn library that is being used in the experiment further.
This Dataet is first converted into a Pandas Dataframe to perform the further analysis.

The  ```  df.head(2)  ```  prints the first 2 tuples of the dataset.

The  ```  df.info()  ```  prints the information of  all the coluns of the dataset.


In [0]:
df

mean radius  mean texture  ...  worst symmetry  worst fractal dimension
0          17.99         10.38  ...          0.4601                  0.11890
1          20.57         17.77  ...          0.2750                  0.08902
2          19.69         21.25  ...          0.3613                  0.08758
3          11.42         20.38  ...          0.6638                  0.17300
4          20.29         14.34  ...          0.2364                  0.07678
..           ...           ...  ...             ...                      ...
564        21.56         22.39  ...          0.2060                  0.07115
565        20.13         28.25  ...          0.2572                  0.06637
566        16.60         28.08  ...          0.2218                  0.07820
567        20.60         29.33  ...          0.4087                  0.12400
568         7.76         24.54  ...          0.2871                  0.07039

[569 rows x 30 columns]

The above snippet prints the dataset. along with size of the entire dataset.

In [0]:
X_train ,  x_test, Y_train, y_test = train_test_split(df, cancer['target'],test_size= 0.3, random_state = 40)
print(X_train, Y_train)

     mean radius  mean texture  ...  worst symmetry  worst fractal dimension
146        11.80         16.58  ...          0.5774                  0.10300
345        10.26         14.71  ...          0.2434                  0.08488
28         15.30         25.27  ...          0.4027                  0.09876
497        12.47         17.31  ...          0.3035                  0.07661
257        15.32         17.27  ...          0.3258                  0.11910
..           ...           ...  ...             ...                      ...
440        10.97         17.20  ...          0.2540                  0.09532
165        14.97         19.76  ...          0.2646                  0.06085
7          13.71         20.83  ...          0.3196                  0.11510
219        19.53         32.47  ...          0.2713                  0.07568
326        14.11         12.88  ...          0.2100                  0.07083

[398 rows x 30 columns] [0 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 0 0 1 0 

The above code splits the input dataset into training set on which the model is being trained and the test dataset on which the model is evaluated. The ```  test_size = 0.3``` specifies that 30% of the dataset will be used to test the model


In [0]:
from sklearn.svm import SVC

In [0]:
svm = SVC()

In [0]:
svm.fit(X_train, Y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

The ``` svm.fit()``` trains the model depending on the input dataset i.e. X_train. The input X_train is the input of the training dataset and Y_train output of the training dataset.

In [0]:
prediction = svm.predict(x_test)

The ``` svm.predict``` predicts the output value for the test dataset.

The output of ``` svm.predict()``` is compared with the actual values i.e. y_test and a report is produced to analyze the results.

In [0]:
from sklearn.metrics import classification_report 

In [0]:
results  = classification_report(y_test, prediction)

In [0]:
print(results)

              precision    recall  f1-score   support

           0       0.94      0.84      0.89        56
           1       0.93      0.97      0.95       115

    accuracy                           0.93       171
   macro avg       0.93      0.91      0.92       171
weighted avg       0.93      0.93      0.93       171



In order to improve the results of the model, GridSearch can be performed.
It takes inputs as C and gamma and decides the values that could produce best results.

In [0]:
from sklearn.model_selection import GridSearchCV

In [0]:
param_grid = {'C':[0.1, 1,10, 100 ], 'gamma':[1,0.1,0.01,0.0001]}

Thus, the list C and gamma are passed to the GridSearch Object .
Also the verbose in the following cells specifies the amount of print statements that are taking place in the process.

In [0]:
grid  = GridSearchCV(SVC(), param_grid,verbose = 3)

In [0]:
grid.fit(X_train, Y_train)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
[CV] C=0.1, gamma=1 ..................................................
[CV] ...................... C=0.1, gamma=1, score=0.600, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ...................... C=0.1, gamma=1, score=0.613, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ...................... C=0.1, gamma=1, score=0.613, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ...................... C=0.1, gamma=1, score=0.608, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ...................... C=0.1, gamma=1, score=0.608, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] .................... C=0.1, gamma=0.1, score=0.600, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ...........

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s


[CV] ................... C=0.1, gamma=0.01, score=0.608, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] ................. C=0.1, gamma=0.0001, score=0.938, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] ................. C=0.1, gamma=0.0001, score=0.912, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] ................. C=0.1, gamma=0.0001, score=0.887, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] ................. C=0.1, gamma=0.0001, score=0.924, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] ................. C=0.1, gamma=0.0001, score=0.886, total=   0.0s
[CV] C=1, gamma=1 ....................................................
[CV] ........................ C=1, gamma=1, score=0.600, total=   0.0s
[CV] C=1, gamma=1 ....................................................
[CV] .

[Parallel(n_jobs=1)]: Done  80 out of  80 | elapsed:    1.0s finished


GridSearchCV(cv=None, error_score=nan,
             estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                           class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='scale', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': [0.1, 1, 10, 100],
                         'gamma': [1, 0.1, 0.01, 0.0001]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=3)

In [0]:
print(grid.best_estimator_)

SVC(C=10, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.0001, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)


The ```best_estimator_``` finds the best value among all the above combinations. 

The ```best_params_``` finds the values of C and gamma for which the models if most accurate. 

In [0]:
print(grid.best_params_) 

{'C': 10, 'gamma': 0.0001}


In [0]:
grid_predictions = grid.predict(x_test)

In [0]:
results  = classification_report(y_test, grid_predictions)

In [0]:
print(results)

              precision    recall  f1-score   support

           0       0.94      0.89      0.92        56
           1       0.95      0.97      0.96       115

    accuracy                           0.95       171
   macro avg       0.95      0.93      0.94       171
weighted avg       0.95      0.95      0.95       171



These are the results obtained when the values of C and gamma are changed as per the output of the grid search. 
It can be seen that the accuracy changes from 93% to 95%.


---
**Conclusion:**
Thus, in the above experiment, I have installed the jupyter notebook and implementted the Support Vector Machine of the sklearn library.
